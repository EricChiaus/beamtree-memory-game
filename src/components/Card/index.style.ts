import styled from 'styled-components';

export const CardWrapper = styled.div`
    width: 100px;
    height: 140px;
    font-size: 64px;
    display: flex;
    justify-content: center;
    align-items: center;
    border: 1px solid black;
    box-sizing: border-box;
`;
