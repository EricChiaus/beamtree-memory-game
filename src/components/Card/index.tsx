import React, { useContext, useMemo } from 'react';
import GameContext from './../../contexts/GameContext';
import { CardWrapper } from './index.style';

interface CardProps {
    index: number;
}

const Card: React.FC<CardProps> = ({ index }) => {
    const { getItem } = useContext(GameContext);

    const item = useMemo<ReturnType<typeof getItem>>(
        () => getItem(index),
        [index, getItem],
    );

    if (!item) {
        return null;
    }

    return <CardWrapper>{item.content}</CardWrapper>;
};

export default Card;
