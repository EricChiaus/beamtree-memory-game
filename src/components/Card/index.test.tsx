import React, { useContext } from 'react';
import { render } from '@testing-library/react';
import Card from '.';

jest.mock('react', () => ({
    ...jest.requireActual('react'),
    useContext: jest.fn(),
}));
const mockUseContext = useContext as jest.Mock;

jest.mock('./index.style', () => ({
    CardWrapper: ({ children }) => {
        return <div>{children}</div>;
    },
}));

describe('<Card />', () => {
    it('should return null given no item found', () => {
        mockUseContext.mockReset().mockReturnValue({
            getItem: () => null,
        });
        const { container } = render(<Card index={0} />);
        expect(container).toMatchInlineSnapshot(`<div />`);
    });

    it('should match the snapshot given valid item', () => {
        mockUseContext.mockReset().mockReturnValue({
            getItem: index => ({
                index: index,
                content: index.toString(),
                isFlippable: true,
                isFlipped: true,
            }),
        });
        const { container } = render(<Card index={0} />);
        expect(container).toMatchSnapshot();
    });
});
