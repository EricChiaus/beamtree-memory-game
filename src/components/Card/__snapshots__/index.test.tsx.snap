// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card /> should match the snapshot given valid item 1`] = `
<div>
  <div>
    0
  </div>
</div>
`;
